Libraries/LibWeb/GamepadAPI/Gamepad.cpp

```diff
@@ -21,4 +149,478 @@ void Gamepad::initialize(JS::Realm& realm)
     Base::initialize(realm);
 }
 
+void Gamepad::visit_edges(Cell::Visitor& visitor)
+{
+    Base::visit_edges(visitor);
+    visitor.visit(m_buttons);
+    visitor.visit(m_vibration_actuator);
+}
+
+void Gamepad::finalize()
+{
+    SDL_CloseGamepad(m_sdl_gamepad);
```

Comment:
We're only closing the SDL resource once the cell gets finalized by GC - is that OK? Or should we be more proactive about this? (e.g. on gamepad disconnect)

Reply:
As far as I can tell it's okay, it'll just stop received updates, plus joystick IDs will never be reused in the applications lifetime. Additionally, SDL_CloseGamepad frees all the memory associated with the gamepad (https://github.com/libsdl-org/SDL/blob/main/src/joystick/SDL_gamepad.c#L3918-L3922), so you could cause a UAF by holding onto the haptic actuator and sending a rumble effect to the dangling `SDL_Gamepad*` when a gamepad is disconnected.